### ISSUE TITLE
Non-panicking versions of functions

### ISSUE THREAD

**Issue by ammkrn** [cessen/ropey#39]
I'd like to know if you would accept a PR adding a feature-gated api extension which includes indexing functions that return `Option<A>` instead of panicking. Thanks!

**Comment by cessen** [comment]
Depending on the design, I don't think it needs to be feature-gated.  Are you thinking something similar to Rust's standard library, where it has separate `get_` variants of the indexing functions?

If so, I'd say go ahead.  I might have some minor bike-shedding on the PR, but in general I'm happy to include that in Ropey.

Also, in order to keep things relatively DRY it probably makes sense to move the main code into the `get_` variants, and then the existing panicking functions can just call those with an `.expect()` on the end, or something along those lines.

**Comment by ammkrn** [comment]
Yeah, that's the idea. I'll defer to you on the naming, for now I went with `try_<name>`. If you're okay with the approach taken in here: https://github.com/ammkrn/ropey/blob/total/src/slice.rs ([diff](https://github.com/cessen/ropey/compare/master...ammkrn:total)), I'll finish the rope module and amend that commit.

Also let me know if you want the docstrings to be the whole thing minus the `panics` section, or just something like `total version of [byte_to_char]`.

**Comment by cessen** [comment]
Oh!  That's a lot more pervasive than I was imagining.  I was thinking of just a few of the most container-like functions:

- The basic getting functions: `byte()`, `char()` and `line()` ->`get_byte()`, `get_char()` and `get_line()`
- The basic mutation functions: `insert()`, `insert_char()` and `remove()` -> `try_insert()`, `try_insert_char()` and `try_remove()`.
- And `slice()` -> `try_slice()`.

Hmm.  I'm hesitant to just go whole-hog on adding alternatives for the entire API, mostly because I expect it will make the documentation a bit too noisy (it's already pretty noisy as-is).  Do you have a concrete use-case for this?  And if so, which functions do you actually expect to need this on?

(Sorry for the whiplash on this.)

**Comment by ammkrn** [comment]
No problem, I just didn't want to do a drive-by along the lines of "here's exactly the methods I need, future users be damned". This branch has the reduced version https://github.com/ammkrn/ropey/tree/total_small.

The concrete use case is this thing: https://github.com/helix-editor/helix. Some of them have been closed, but many of the initial issues being filed are panics due to off-by-n errors while working with Rope and RopeSlice.

The only new methods in the total_small branch are the ones directly used by helix, which are:
```
RopeSlice:
  slice
  char
  char_to_line
  line_to_char
  line
  chars_at
  char_to_byte
  chunk_at_byte
  byte_to_char

Rope:
  slice
  remove
  insert
  char_to_byte
  char_to_line
```

Rope.slice(), Rope.remove(), and RopeSlice.slice() do contain duplicate logic as a trade-off in order to preserve the nice error reporting without adding a clone requirement on the RangeBounds type, or calculating `start, end` upfront. These are obviously really cheap, so I'm happy to change it to do just that; please let me know what you think is more appropriate.

**Comment by cessen** [comment]
> The concrete use case is this thing: https://github.com/helix-editor/helix. Some of them have been closed, but many of the initial issues being filed are panics due to off-by-n errors while working with Rope and RopeSlice.

Hmm.  I suspect that these aren't generally recoverable errors.  For example, if you're trying to insert a character off the end of the rope, or convert character positions that don't exist, that indicates a serious bug that needs to be fixed anyway.  The use-case I was imagining was a situation where you actually want to call these as part of correct logic, where you expect the functions to sometimes fail in the course of normal bug-free operation.

Is the goal in your case to fail gracefully (e.g. save the file(s) to a backup before crashing cleanly) rather than panic?

I don't mean to push back *too* hard on this.  But I think this may be a mismatch of use-case.  The general approach in Rust (as far as I understand) is that things like `Option` and `Result` are for expected control flow, rather than bugs.  And panics are precisely for the when something buggy happens.

**Comment by ammkrn** [comment]
I agree with the last point, but I think the issue of "there are unrecoverable off-by-n bugs in the core editor" and "the core editor should not be using partial functions to index collections" are separable, and the latter is going to persist after the former is fixed. For modern editors that integrate with external tools like language servers and allow users to write plugins/extensions that can control things like the buffer contents, selections, etc., the author(s) of the core editor can't necessarily be expected to know that they're always going to be indexing in-bounds since they're relinquishing a great deal of control. 
I only heard about helix two days ago so I don't know the code base well, but the author did agree that "a safe indexing wrapper would be useful" when the issue was brought up (https://github.com/helix-editor/helix/issues/43#issuecomment-852710325), which led me to believe the potential for OOB rope accesses would remain even after the initial bugs were worked out, and that the ability to recover would be in some way helpful. I know I already mentioned this, but feature-gating the extended methods would preserve the current API.

**Comment by cessen** [comment]
> allow users to write plugins/extensions that can control things like the buffer contents, selections, etc.,

Ah, that's an excellent point.

Thanks for the link to the issue on helix's repo.

**Comment by cessen** [comment]
Okay, so I basically have two conflicting concerns here:

1. I think this would be legitimately useful functionality, and I'd like it to be in Ropey.
2. I don't want it making a confusing mess of the documentation.

Possible solutions considered so far:

1. Only include non-panicking variants for a small subset of functions.  IMO this only makes sense if there's a clear, conceptually clean subset of functions that's small enough to address the concrete use-cases.  And I suspect that's probably not the case.
2. Feature gate the new API.  To me, it feels weird to feature gate something just to keep the documentation clean.  Ideally, I'd like this to just be available by default, because it doesn't seem special enough to be a feature.

So neither of those feel like good solutions to me.  But here are a couple of other possibilities:

1. Create a trait for the functionality, which can be optionally imported by the client code.  This would put the non-panicking versions of everything in their own area of the documentation, with a link from the main types to make it discoverable.
2. Just add a second `impl` block to `Rope` and `RopeSlice` and put the new functions in there.  IIRC that puts them in their own collapsible block in the documentation, which would make things easier to browse.

I'm going to play around with both of these, and see which one feels better to me.

**Comment by archseer** [comment]
I think one case where it would make sense to add a safe wrapper is the `char` function. In many situations I needed to fetch the char at cursor plus the next char: 

```
        let ch = slice.char(begin);
        let next = slice.char(begin + 1);
```

I already know that `begin` is in bounds since it's at cursor location, but `next` might not be. I use this wrapper in some parts, but there was older code that didn't do the bounds check, hence the panics.

```rust
fn next_char(doc: &Rope, pos: usize) -> Option<char> {
    if pos >= doc.len_chars() {
        return None;
    }
    Some(doc.char(pos))
}
```

**Comment by cessen** [comment]
Hi @archseer!

I think at this point, I'm planning to just have non-panicking versions of all the methods.  It's mostly a matter of how to integrate that in a way that doesn't make documentation too cluttered and hard to navigate (which it already is to some extent, even now).

After playing around with the two alternatives I mentioned above, I think it probably makes the most sense to just put them in their own `impl` block, just so they're not interwoven with all the other methods.  That approach also gives us plenty of room to just generally organize and re-organize the documentation via `impl` blocks in the future without breaking API compatibility.

I do have one last question I'm trying to figure out, though: would it make sense for some of these functions to return a `Result` instead of an `Option`?

For a function like `get_char()`, returning an `Option` is pretty obviously the way to go, I think.  But maybe the index conversion functions like `try_byte_to_char()` should return a `Result` instead?  Conceptually it feels to me like functions that "fetch" something feel more `Option`-like to me, whereas failed computations like index conversion feel more error-like.  But maybe I'm over-thinking this.

**Comment by cessen** [comment]
@ammkrn If you still want to take a stab at this, please go ahead!  Although if you'd rather not after the run-around I've given you guys, I totally understand (and in which case I'll do it myself).

At this point I'm committed to providing these API's.  It's mostly just a matter of bike-shedding at this point.  How things are documented, organized, named, and whether we use `Option` or `Result` isn't critical for a first implementation and PR merge.  They're just things we'll want to figure out before committing to the specifics of the API in a release.

**Comment by ammkrn** [comment]
It's all good, I'll take another look at it tonight. I might rediscover cases in which `Result` is more appropriate, but I would generally lean toward `Option` since that adheres more closely to the style used in std ala `Vec::get()`.

**Comment by cessen** [comment]
> but I would generally lean toward `Option` since that adheres more closely to the style used in std ala `Vec::get()`.

Right, for functions that are obviously `get()`-like (e.g. `byte()`, `char()`, etc.) `Option` is definitely the way to go.  It's everything else that I'm not as sure about.  For example, `byte_to_char()` and the other index-conversion methods, which don't have much equivalent in the `std` APIs as far as I'm aware.

But don't worry about it too much.  I'll do my own pass after merging the PR.

Thanks so much for your time and contributions!

**Comment by cessen** [comment]
Resolved by #42 (thanks to @ammkrn!)

There's still a bit of bike shedding and a documentation pass I'd like to do, but I'll take care of that piece-meal the next time I start gearing up for a release.  Regardless, the functionality is in! :-)
